model/config.go

```diff
@@ -1993,6 +2030,47 @@ func (ls *LocalizationSettings) isValid() *AppError {
 	return nil
 }
 
+func (mes *MessageExportSettings) isValid(fs FileSettings) *AppError {
+	if mes.EnableExport == nil {
+		return NewAppError("Config.IsValid", "model.config.is_valid.message_export.enable.app_error", nil, "", http.StatusBadRequest)
+	} else if *mes.EnableExport {
+		if mes.ExportFromTimestamp == nil || *mes.ExportFromTimestamp < 0 || *mes.ExportFromTimestamp > time.Now().Unix() {
+			return NewAppError("Config.IsValid", "model.config.is_valid.message_export.export_from.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if mes.DailyRunTime == nil {
+			return NewAppError("Config.IsValid", "model.config.is_valid.message_export.daily_runtime.app_error", nil, "", http.StatusBadRequest)
+		} else if _, err := time.Parse("15:04", *mes.DailyRunTime); err != nil {
+			return NewAppError("Config.IsValid", "model.config.is_valid.message_export.daily_runtime.app_error", nil, err.Error(), http.StatusBadRequest)
+		}
+
+		if mes.FileLocation == nil {
+			return NewAppError("Config.IsValid", "model.config.is_valid.message_export.file_location.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if !(*fs.DriverName == IMAGE_DRIVER_LOCAL) {
+			var absFileDir, absMessageExportDir string
+			var err error
+			if absFileDir, err = filepath.Abs(fs.Directory); err != nil {
+				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.file_location.relative", nil, err.Error(), http.StatusBadRequest)
+			}
+			if absMessageExportDir, err = filepath.Abs(*mes.FileLocation); err != nil {
+				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.file_location.relative", nil, err.Error(), http.StatusBadRequest)
+			}
+			if !strings.HasPrefix(absMessageExportDir, absFileDir) {
+				// configured export directory must be relative to data directory
+				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.file_location.relative", nil, "", http.StatusBadRequest)
+			}
```

Comment:
In case you weren't aware, variables declared in `if` statements are accessible from `else if` blocks:

```go
if absFileDir, err := filepath.Abs(fs.Directory); err != nil {
} else if absMessageExportDir, err := filepath.Abs(*mes.FileLocation); err != nil {
} else if !strings.HasPrefix(absMessageExportDir, absFileDir) {
}
```

So you could cut out several lines by doing that.

Reply:
Good tip. I didn't think to do that, but it looks very natural with Golang's if-based error handling syntax.